MAINT Create private `_pairwise_distances_reductions` submodule

jjerphan opened the pull request:
#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Follow-up of scikit-learn#23515.

#### What does this implement/fix? Explain your changes.

The single source file is split in several files to create a private submodule:
 - an extension per file ([similarly to the `tree` submodule](https://github.com/scikit-learn/scikit-learn/blob/70eebb992e8a799cbc3d7f1fbfddd104c0908c66/sklearn/tree/setup.py))
 - dispatchers are currently present with their associated implementation so as to be able to template implementation using Tempita
 - import paths have not changed for Python callers

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jjerphan committed:
MAINT Introduce interfaces for PairwiseDistancesReductions

Those interfaces are meant to be used in the Python code, decoupling the
actual implementation from the Python code. This allows changing all the
private implementation while maintaining a contract for the Python
callers.

Each interface extending the base `PairwiseDistancesReduction` interface
must implement the :meth:`compute` classmethod.

Under the hood, such a function must only define the logic to dispatch
at runtime to the correct dtype-specialized `PairwiseDistancesReduction`
implementation based on the dtype of X and of Y.

This refactoring will ease other dtype support such as float32 support.

jjerphan committed:
DEBUG Propagate sort_results

jjerphan committed:
DOC Reword

jjerphan committed:
DOC Document dispatchers and implementations

… with some ASCII art. 🎨

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Merge branch 'main' into maint/pairwise-distances-reductions-interfaces

jjerphan committed:
MAINT Create private _pairwise_distances_reductions submodule

jjerphan committed:
MAINT Simply use python class for dispatchers

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan committed:
MAINT Group dispatchers in a single file

jjerphan committed:
DOC Improve comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Apply review comments

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
DOC Fix typo

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

jjerphan committed:
DOC Move comment where appropriate

jjerphan committed:
MAINT Dispatch _sqeuclidean_row_norms

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Merge branch 'main' into maint/pdr-private-submodule

jjerphan committed:
Correctly handle read-only datasets

Also reword comments in tests

jjerphan committed:
Apply review comment

From: https://github.com/jjerphan/scikit-learn/pull/14#pullrequestreview-1004501769

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jeremiedbb commented on sklearn/metrics/_pairwise_distances_reduction/__init__.py:
```diff
@@ -0,0 +1,101 @@
+    PairwiseDistancesReduction,
+    PairwiseDistancesArgKmin,
+    PairwiseDistancesRadiusNeighborhood,
+    _sqeuclidean_row_norms,
+)
+
+__all__ = [
+    "PairwiseDistancesReduction",
+    "PairwiseDistancesArgKmin",
+    "PairwiseDistancesRadiusNeighborhood",
+    "_sqeuclidean_row_norms",
```
I think we can define ``sqeuclidean_row_norms`` without the prefix underscore, like the dispacher classes.

jeremiedbb commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -0,0 +1,343 @@
+
+
+def _sqeuclidean_row_norms(X, num_threads):
+    """Compute the squared euclidean norm of the rows of X in parallel."""
```
Probably worth adding that X is expected to be C ordered

jeremiedbb commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -0,0 +1,343 @@
+    )
+
+
+class PairwiseDistancesReduction:
+    """Abstract base dispatcher for pairwise distance computation & reduction.
+
+    Each dispatcher extending the base :class:`PairwiseDistancesReduction`
+    dispatcher must implement the :meth:`compute` classmethod.
+    """
```
The lost the definition of ``compute`` as an abstract class method in the transfer

jeremiedbb reviewed (commented):
I like the new organization. A lot more readable !

What's the purpose of having an empty ``test`` directory inside the ``_pairwise_distances_reduction`` sub-module ?

jjerphan committed:
Apply review comments

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan commented:
> I like the new organization. A lot more readable !

Thanks for the review. It definitely was time to.

> What's the purpose of having an empty test directory inside the _pairwise_distances_reduction sub-module ?

Just a re-factoring left-over.

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -0,0 +1,385 @@
+        -------
+        True if the PairwiseDistancesReduction can be used, else False.
+        """
+        dtypes_validity = X.dtype == Y.dtype and Y.dtype == np.float64
```
Can be simplified to:

```suggestion
        dtypes_validity = X.dtype == Y.dtype == np.float64
```

ogrisel reviewed (approved):
LGTM!

jjerphan committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan commented on sklearn/metrics/tests/test_pairwise_distances_reduction.py:
```diff
@@ -952,6 +953,58 @@ def test_pairwise_distances_radius_neighbors(
+    dtype=np.float64,
+):
+    # Results must not depend on the datasets writability
+    rng = np.random.RandomState(global_random_seed)
```
Is there a commit that test `global_random_seed` for all seeds on the CI using https://github.com/scikit-learn/scikit-learn/pull/23026#issue-1190347858 ?

Also, does this test require `global_random_seed`?

jjerphan replied:
> Is there a commit that test global_random_seed for all seeds on the CI using #23026 (comment) ?

No, thus I've pushed dc26da073615b309fd953045d52ddef9e6effa74.

> Also, does this test require global_random_seed?

No, removed in 94c76fbaa7f9169abba8cf12eb6c42c84bc2d550.

thomasjpfan replied:
I pushed [`b089de3` (#23724)](https://github.com/scikit-learn/scikit-learn/pull/23724/commits/b089de367b50a0e4c905745d58e3eb21604faeeb) to trigger the random seeds. [`dc26da0` (#23724)](https://github.com/scikit-learn/scikit-learn/pull/23724/commits/dc26da073615b309fd953045d52ddef9e6effa74) failed because of the "Co-authored-..." in the message.

thomasjpfan reviewed (commented):
Thank you for the PR! I like the refactor.

jjerphan committed:
Merge branch 'main' into maint/pdr-private-submodule

jjerphan committed:
TST Remove useless globam random seed

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
CI Trigger some tests on supporting global random seed [all random seeds]

test_chunk_size_agnosticism
test_n_threads_agnosticism
test_strategies_consistency
test_pairwise_distances_argkmin
test_pairwise_distances_radius_neighbors
test_sqeuclidean_row_norms

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan committed:
CI Trigger [all random seeds]
test_chunk_size_agnosticism
test_n_threads_agnosticism
test_strategies_consistency
test_pairwise_distances_argkmin
test_pairwise_distances_radius_neighbors
test_sqeuclidean_row_norms

thomasjpfan committed:
Merge branch 'main' into maint/pdr-private-submodule

thomasjpfan reviewed (approved):
LGTM
